File: sources/02-abstract-data-types/03-queue/c/Main.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Código fonte - Implementação de Fila
 *
 * @author Artur Bomtempo Colen
 * @algorithm Data Structure
 * @type Queue
 * @language C
 * @version 1.0
 */

#define LENGTH 6

int array[LENGTH];
int front = -1;
int rear = -1;

void start() { front = rear = -1; }

int isFull() { return rear == LENGTH - 1; }

int isEmpty() { return front == -1; }

void enqueue(int value) {
    if (isFull()) {
        printf("Erro: Fila cheia.\n");
        exit(1);
    }

    if (front == -1) {
        front = 0;
    }

    rear++;
    array[rear] = value;
}

int dequeue() {
    if (isEmpty()) {
        printf("Erro: Fila vazia.\n");
        exit(1);
    }

    int value = array[front];

    if (front == rear) {
        front = rear = -1;
    } else {
        front++;
    }

    return value;
}

void show() {
    if (isEmpty()) {
        printf("Fila vazia.\n");
        return;
    }

    printf("Fila: [");

    for (int i = front; i <= rear; i++) {
        printf("%d", array[i]);

        if (i < rear) {
            printf(", ");
        }
    }

    printf("]\n");
}

int main() {
    printf("===== Fila =====\n");

    start();

    enqueue(1);
    enqueue(7);
    enqueue(9);
    enqueue(3);
    enqueue(8);
    enqueue(4);

    show();

    int x1 = dequeue();
    int x2 = dequeue();
    int x3 = dequeue();

    printf("%d, %d, %d\n", x1, x2, x3);

    show();

    return 0;
}

File: sources/02-abstract-data-types/03-queue/java/Main.java
/**
 * Código fonte - Implementação de Fila
 *
 * @author Artur Bomtempo Colen
 * @algorithm Data Structure
 * @type Queue
 * @language Java
 * @version 1.0
 */

class Queue {
    private static final int LENGTH = 6;
    private static int[] array = new int[LENGTH];
    private static int front = -1;
    private static int rear = -1;

    public static void start() {
        front = rear = -1;
    }

    public static boolean isFull() {
        return rear == LENGTH - 1;
    }

    public static boolean isEmpty() {
        return front == -1;
    }

    public static void enqueue(int value) {
        if (isFull()) {
            System.out.println("Erro: Fila cheia.");
            System.exit(1);
        }

        if (front == -1) {
            front = 0;
        }

        rear++;
        array[rear] = value;
    }

    public static int dequeue() {
        if (isEmpty()) {
            System.out.println("Erro: Fila vazia.");
            System.exit(1);
        }

        int value = array[front];

        if (front == rear) {
            front = rear = -1;
        } else {
            front++;
        }

        return value;
    }

    public static void show() {
        if (isEmpty()) {
            System.out.println("Fila vazia.");
            return;
        }

        System.out.print("Fila: [");

        for (int i = front; i <= rear; i++) {
            System.out.print(array[i]);

            if (i < rear) {
                System.out.print(", ");
            }
        }

        System.out.println("]");
    }
}

public class Main {
    public static void main(String[] args) {
        System.out.println("===== Fila =====");

        Queue.start();

        Queue.enqueue(1);
        Queue.enqueue(7);
        Queue.enqueue(9);
        Queue.enqueue(3);
        Queue.enqueue(8);
        Queue.enqueue(4);

        Queue.show();

        int x1 = Queue.dequeue();
        int x2 = Queue.dequeue();
        int x3 = Queue.dequeue();

        System.out.println(x1 + ", " + x2 + ", " + x3);

        Queue.show();
    }
}